### ISSUE TITLE
Conflicting variable names in deal.II and Qt

### ISSUE THREAD

**Issue by philm001** [dealii/dealii#7524]
Hello all,

I recently came across a bug where a variable in the global namespace is conflicting with a define in the QT library. In grid/tria.h line 2335, we have 

`mutable Signals signals `

This is conflicting with a Qt library define statement found in QtCore/qobjectdefs.h line 93 where the statement is 
`#define signals Q_SIGNALS`

I thought that I would bring this to your attention. I am not sure where I can submit issues with the Qt library so I thought that I could at least give you guys a heads up.

**Comment by Rombur** [comment]
@philm001 `mutable Signals signals` is not define in the global namespace it is in the `dealii` namespace but macros ignore the namespace unfortunately.

**Comment by masterleinad** [comment]
In other words: `deal.II` is not responsible for the problem, but `Qt` is. We were having a similar issue with the macros `Assert`, `AssertThrow` and the like, but recently implemented a possibility to disable these macros, see #7492. Maybe, there exists a similar switch in `Qt`. Otherwise, you can just try to undefine the `Qt` macro after including all the headers if it is not really used in your code.

**Comment by masterleinad** [comment]
`Qt` issues can be submitted at https://bugreports.qt.io/secure/CreateIssue!default.jspa.

**Comment by drwells** [comment]
Its not great that we have this conflict (particularly with a popular library like QT), but that macro has been in QT since the 1990s and isn't going anywhere (i.e., signals and slots are [a primary feature of QT](http://doc.qt.io/archives/qt-4.8/signalsandslots.html)).

I don't think we can really address this since `Triangulation::signals` is public and the issue comes from poor macro hygiene in another library.

**Comment by bangerth** [comment]
We could emit a warning in this file if the macro has been defined before. Something like this maybe?
```
#ifdef signals
#  warning ...
#endif
```
Somewhere immediately before the definition of the variable (or the top of the file)?

**Comment by philm001** [comment]
I can confirm that the signals in Qt are used to process UI events. So I don't think that it is a good idea to undefine the signals. I have used Qt before on another project that I was working on.

I haven't thoroughly gone through the source code on Deal.II yet (Still working on that one) but what is the possibility of renaming the macro?

To solve the issue on my end, I am thinking to use the refactoring tool in eclipse to rename the macro to:

mutable Signals dealii_signals.

**Comment by bangerth** [comment]
The way to work around this is by `#include`ing the Qt headers *after* the deal.II headers.

**Comment by philm001** [comment]
Thank you. I am not at my computer right now but I will give it a shot as soon as possible

**Comment by philm001** [comment]
So here is what is happening, Eclipse is deciding to compile the files containing the Qt includes first. Then it compiles the dealii code. There are no Qt includes in the dealii code. The two sets of code and separate from each other.

I am looking into the possibility of changing the build order of the files for eclipse. But, if this doesn't pan out, I am still entertaining the idea of changing all instances of the variable name.

**Comment by bangerth** [comment]
This does not seem correct. Each `.cc` file is compiled separately. The *order* in which the `.cc` files are compiled does not matter. What does matter is the order in which `.h` files are `#include`d in the `.cc` files. So that's where you need to deal with the problem.

**Comment by philm001** [comment]
Yes, I was looking at the build order for the `.h ` files. I apologize that my previous post was not very clear.

Anyways, I felt that this method only provided a band-aid solution for the problem at hand. I ended up changing the name using the refactor tool in eclipse and I was able to successfully recompiled the Deal II library. I am working on the stable version 9.0.1. I later recompiled my application and it was able to build without any errors.

If you are interested, I can download a copy of the development branch and use the same tool to rename the macro and submit a pull request. All I would need is a new name for the macro. I named my version to be `tri_signals`.

**Comment by masterleinad** [comment]
> I don't think we can really address this since Triangulation::signals is public and the issue comes from poor macro hygiene in another library.

Changing the name of the variable directly is not backward-compatible. The only thing we could do is to define the old one to be a reference to the new one and deprecate it such that it will be removed in about a year. That, however, implies that the old name is still around for that time. Hence, this solution doesn't really help in the short-run.
In the end, we can't fix all the problems in our code base other libraries are responsible for.

> I can confirm that the signals in Qt are used to process UI events. So I don't think that it is a good idea to undefine the signals. I have used Qt before on another project that I was working on.

If you just undefine the macro after including all the `Qt` headers, you should be fine if you just replace all occurrences of `Qt`'s `signals` by ` Q_SIGNALS` in your code. Alternatively, @bangerth's solution should work as well. Both of these solutions don't require changes in `deal.II` and it is much easier to update to a newer Release or to keep track with the developer version.

**Comment by philm001** [comment]
Hey guys, just wanted to update you on this. I decided to change the order of my #include statements as outlined by Dr. Bangerth and was able to get the application compiled. 

After doing some further research in Qt, there is a define statement that a developer can use that tells Qt to not define the MOC keyword signal (and other keywords). But to instead use the Q_SIGNAL macro. Here is a link in regards to this: https://doc.qt.io/qt-5/signalsandslots.html#using-qt-with-3rd-party-signals-and-slots

For those not using qt creator, use the preprocessor macro QT_NO_KEYWORDS.

I remember it took me a little bit to figure the bug out and to boil it down. I was thinking that it might be a good idea to at least output a warning (or error) message that tells the developer if the signals macro is already defined somewhere. And to maybe give some ideas on fixing it.

I haven't kept up with the latest releases so if this isn't already there, I can look into coding it?

**Comment by masterleinad** [comment]
We didn't change anything in this regard in `deal.II`.
If you want to come up with a way to warn the user, we are happy to discuss your solution. You probably want to add something in `include/deal.II/base/config.h`.

**Comment by philm001** [comment]
Yes, I would like to come up with a solution to warn the user. I will take a look at the file and think of some solutions and continue to post in this thread

**Comment by bangerth** [comment]
@philm001 -- any updates on this?

**Comment by philm001** [comment]
@bangerth Hey guys sorry for disappearing on this! My life got crazy for awhile and I had to drop alot of the things that I was doing. But since quarantine life kicked, I have a lot of free on my hands. A better update will be coming soon-ish on this. Currently need to update my copy of dealii. It is a little out of date

**Comment by philm001** [comment]
@masterleinad Hello, I have pulled the latest version of dealii. When you state config.h, do you mean the config.h.in?

is this file meant for define statement related to building Dealii?

**Comment by drwells** [comment]
config.h is the configuration header generated by CMake - if your build directory is named `build`, then it is `build/include/deal.II/base/config.h`.

**Comment by philm001** [comment]
So if config.h is a file generated by cmake, I don't think that placing the warning in config.h is a good idea since this file is a cmake processed file and is local to the user's build.  I would want to place in the library perhaps? config.h.in is a good contender since that file becomes config.h after running cmake. I can place the warning there.

My next option on where to place would be tri.h (include/deal.ii/grid) And place it right above the `mutable Signals signals` line. But it wouldn't match the formatting of the file. So I don't think that it is a good idea.

**Comment by masterleinad** [comment]
I think having it at the top of `include/deal.II/grid/tria.h` and `include/deal.II/multigrid/multigrid.h` is suitable.

**Comment by philm001** [comment]
@masterleinad ok, I will place it there

**Comment by philm001** [comment]
ok, I have made the modifications, I just need to run some tests and issue a pull request. Here is the message that I wrote. Let me know if you want to change it:

The data type signals is already defined. You are most likely using the QT library \
and using the signals keyword. You can either #include the Qt headers (or any conflicting headers) \
after the deal.ii headers or you can define the QT_NO_KEYWORDS marco and use the Q_SIGNALS macro."

**Comment by philm001** [comment]
Hey guys, I ran some tests and everything works out! I was able to get the error message when compiling under Qt Creator. When I switched the headers around, the program compiled successfully.

All I need to do is to update my version with the latest. I think that there have been a number of updates since I wrote the code. 

Let me know if my error message is good. I will be submitting a pull request soon (will need to figure out how to update a repo within github)
